### PR TITLE
[5.4] Add ability to create depth collections

### PR DESCRIPTION
**Pull request by lcjury**
A friend needed the ability to create depth collections, if an array have nested arrays, all of those array's should be collected.

### CODE REVIEW

File: src/Illuminate/Support/Collection.php [hunk 1]
```diff
@@ -59,6 +59,39 @@ public static function make($items = [])
...
+
+    /**
+     * Recursive iteration over arrayable mapping arrays to collections.
+     *
+     * @param mixed $item
+     * @return static
+     */
+    private static function makeDepthRecursive($items)
+    {
+        $items = collect()->getArrayableItems($items);
```
Comment (lcjury) [inline src/Illuminate/Support/Collection.php]: I don't like this, but didn't know how to call `getArrayableItems` without having a working instance of a collection.

Comment (jmarcher) [inline src/Illuminate/Support/Collection.php]: You can use self::getArrayableItems

File: src/Illuminate/Support/Collection.php [hunk 2]
```diff
@@ -59,6 +59,39 @@ public static function make($items = [])
...
+
+        return array_map(function ($item) {
+            if (! is_array($item)) {
+                return $item;
+            }
+            $item = self::makeDepthRecursive($item);
```
Comment (jmarcher) [inline src/Illuminate/Support/Collection.php]: I don't like the idea of using recursion inside an array_map

File: src/Illuminate/Support/helpers.php [hunk 3]
```diff
@@ -379,6 +379,19 @@ function collect($value = null)
...
 }
 
+if (! function_exists('collectDepth')) {
+    /**
+     * Create a collection from a multi-level array.
+     *
+     * @param  mixed  $value
+     * @return \Illuminate\Support\Collection
+     */
+    function collectDepth($value = null)
```
Comment (jmarcher) [inline src/Illuminate/Support/helpers.php]: I would not create a helper function for a functionality that we do not know if is going to need a helper function.
And for helper functions we use the slug notation

**Comment by taylorotwell** [comment]
Not adding this to core at this time.

**Pull request closed without merging**